README.md
npx express --git backend

git remote add origin https://github.com/otmjka/pplmap-backend.git
https://gist.github.com/subfuzion/08c5d85437d5d4f00e58

===

git push https://otmjka@github.com/otmjka/pplmap-backend.git

===

curl -d '{"name":"dima", "birthday":495136800000}' -H "Content-Type: application/json" -X POST localhost:3001/persons/add --verbose

1. deploy app to heroku

### benefits wished

- free for try
- postgres service
- automation, deploy by one command

steps:

- build ts app to node version
- try in docker
- build app docker try run
- how to set env variables

## typescript

build js

```
npx tsc --noEmit false
```

```
ENVIRONMENT=local PORT=3001 POSTGRES_URI='postgres://postgres:postgres@127.0.0.1:5432/postgres' node build/index.js
```

## docker

-[heroku docker](https://docs.docker.com/engine/reference/builder/)

```
ENV PORT=3001
ENV POSTGRES_URI=postgres://postgres:postgres@host.docker.internal:5432/postgres
ENV ENVIRONMENT=production
ENV LOG_LEVEL=debug

docker build -t pplmap-backend-000:latest .

// ENV POSTGRES_URI=postgres://postgres:postgres@host.docker.internal:5432/postgres
```

```bash
docker run --rm -it -e DATABASE_URL='host.docker.internal:5432' -e ENVIRONMENT=local -e PORT=3001 -e DB_NAME=postgres -e DB_PASSWORD=postgres -e DB_USER=postgres pplmap-backend-000:latest /bin/bash
```

## docker get access to localhost postgres on mac

- [docker get access to localhost postgres on mac](https://docs.docker.com/docker-for-mac/networking/#i-want-to-connect-from-a-container-to-a-service-on-the-host)

## heroku

### push to heroku

```bash
docker build -t pplmap-backend-000:latest .
heroku container:push web

```

- [setup postgres service](https://devcenter.heroku.com/articles/heroku-postgresql#provisioning-heroku-postgres)
- [heroku docker](https://devcenter.heroku.com/articles/container-registry-and-runtime)


CHANGELOG.md
0. [+] move this list to CHANGELOG.md
1. [+] simple implementation of
   POST /person/add
   GET /persons
   GET /person
2. [+] tests

3. [+] docker
4. [+] pg-promise postgresql
5. [+] deploy to heroku
6. [+] add typescript

?7. swagger
?8. apollo
?9. auth
?10. acme project config
?11. issue with

12. get database data everyday and save to dropbox/google drive
13. try migration
14. pass to container `environment variables` instead `Dockerfile's` `ENV`
15. add


tests/010.test.js
const axios = require('axios')
const appUrl = 'http://localhost:3001'

it('return empty persons', async () => {
  expect.assertions(2)
  await axios.post(`${appUrl}/persons/flush`)
  const result = await axios.get(`${appUrl}/persons`)
  expect(result.status).toBe(200)
  expect(result.data).toHaveLength(0)
})

it('add person', async () => {
  expect.assertions(3)
  const result = await axios.post(`${appUrl}/persons/add`, {name: 'dima', birthday: 495136800000})
  expect(result.status).toBe(200)
  const personsResponse = await axios.get(`${appUrl}/persons`)
  expect(personsResponse.status).toBe(200)
  expect(personsResponse.data).toHaveLength(1)
})


app.js
var createError = require('http-errors');
var express = require('express');
const cors = require('cors');
var path = require('path');
var cookieParser = require('cookie-parser');
var logger = require('morgan');
var {v4: uuidv4} = require('uuid');

var indexRouter = require('./routes/index');
var usersRouter = require('./routes/users');
let persons = []
var app = express();

app.use(cors())
// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'jade');

app.use(logger('dev'));
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));

app.use('/', indexRouter);
app.use('/users', usersRouter);
app.post('/persons/add', (req, res) => {
  persons.push({...req.body, id: uuidv4()})
  res.sendStatus(200)
})

app.post('/persons/flush', (req, res) => {
  persons = []
  res.sendStatus(200)
})

app.get('/persons', (req, res) => {
  res.send(persons)
})

app.get('/persons/:uuid', (req, res, next) => {
  const person = persons.find(p => p === params.uuid)
  if (!person) {
    next(createError(500))
    return
  }
  res.send(person)
})

// catch 404 and forward to error handler
app.use(function(req, res, next) {
  next(createError(404));
});

// error handler
app.use(function(err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.render('error');
});

module.exports = app;


src/Db.ts
import path from 'path';
import { EventEmitter } from 'events';
import pgPromise, { IDatabase, QueryFile } from 'pg-promise';

const QUERY_NAMES = ['personsList', 'addPerson', 'flushPersons'];
const QUERY_FILES_DIR = path.resolve(__dirname, 'sql');

export default class Db extends EventEmitter {
  db: IDatabase<{}>;
  queryFiles: { [queryName: string]: QueryFile };

  constructor({
    databaseUrl,
  }: {
    databaseUrl: string;
  }) {
    super();
    const connectionString = databaseUrl; // `postgres://${dbUser}:${dbPassword}@${databaseUrl}/${dbName}`;
    this.db = pgPromise()(connectionString);

    this.queryFiles = {};
    this.loadQueryFiles();
  }

  loadQueryFiles(): void {
    QUERY_NAMES.forEach((queryName: string) => {
      const queryFile = new QueryFile(
        path.resolve(QUERY_FILES_DIR, `${queryName}.sql`),
      );
      this.queryFiles[queryName] = queryFile;
    });
  }

  async addPerson({
    person_name,
    birthday,
  }: {
    person_name: string;
    birthday: number;
  }): Promise<void> {
    await this.db.none(this.queryFiles.addPerson, {
      person_name,
      birthday: new Date(birthday),
    });
  }

  async personsList({
    skip,
    limit,
  }: {
    skip: number;
    limit: number;
  }): Promise<Array<{ id: string; person_name: string; bithday: Date }>> {
    return await this.db.manyOrNone<{
      id: string;
      person_name: string;
      bithday: Date;
    }>(this.queryFiles.personsList, {
      skip,
      limit,
    });
  }

  async flushPersons(): Promise<void> {
    await this.db.none(this.queryFiles.flushPersons, {});
  }
}


src/index.ts
import Db from './Db';
import HttpServer from './HttpServer';
import getConfig from './config';
import migrate from './migrate/migrate';

let config;

try {
  config = getConfig();
} catch (error) {
  // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-member-access
  console.error({ msg: 'misconfiguration', error_message: error.message });
  process.exit(1);
}
if (!config) {
  // eslint-disable-next-line no-console
  console.error({ msg: 'misconfiguration', error_message: 'empty config' });
  process.exit(1);
}

console.log(
  '###',
  `
  ENVIRONMENT: ${process.env.ENVIRONMENT}
  PORT: ${process.env.PORT}
  DB_NAME: ${process.env.DB_NAME}
  DB_PASSWORD: ${process.env.DB_PASSWORD}
  DB_USER: ${process.env.DB_USER}
  DATABASE_URL: ${process.env.DATABASE_URL}
`,
);

const db = new Db({
  databaseUrl: config.databaseUrl,
});

const httpServer = new HttpServer({
  port: config.port,
  db,
});

(async (): Promise<void> => {
  try {
    console.log(process.cwd());
    await migrate({
      target: '001',
      pgpdb: db.db,
      migrationsPath: `${process.cwd()}/src/sql/migrations`,
    });
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error({
      msg: 'db_init_error',
      error_message: 'error while migration or create schema',
    });
    process.exit(1);
  }
  await httpServer.start();
})();


src/migrate/migrate.ts
import { IDatabase } from 'pg-promise';
import pTimeout from 'p-timeout';

import Db from './Db';
import MigrationFiles from './MigrationFiles';

const MIGRATION_STEP_TIMEOUT_MS = 30 * 1000;

export default async ({
  pgpdb,
  schemaName = 'public',
  migrationsPath,
  target,
}: {
  pgpdb: IDatabase<Record<string, unknown>>;
  schemaName?: string;
  migrationsPath: string;
  target: string;
}): Promise<void> => {
  const db = new Db({ pgpdb, schemaName });
  const migrationFiles = new MigrationFiles(migrationsPath);
  const targetVersion =
    target === 'latest' ? migrationFiles.latestVersion : target;
  console.info({ msg: 'migration_start', targetVersion, schemaName });
  const dbEmpty = await db.isDatabaseEmpty();
  if (dbEmpty) {
    console.info({ msg: 'initializing_empty_database' });
    await db.createSchema();
  }
  const currentVersion = await db.getLastVersion();
  console.info({ msg: 'current_db_version', currentVersion });

  if (currentVersion === targetVersion) {
    console.info({ msg: 'no_migration_needed' });
    return;
  }

  if (targetVersion <= currentVersion) {
    console.error({
      msg: 'required_version_is_lower_than_current',
      targetVersion,
      currentVersion,
    });
    throw new Error('required_db_migration_version_is_lower_than_current');
  }

  const fileChain = migrationFiles.paths[currentVersion][targetVersion];
  if (!fileChain) {
    console.error({
      msg: 'impossible_migration',
      currentVersion,
      targetVersion,
      message:
        'Could not find a query chain to get from current version to target version',
    });
    throw new Error('Impossible migration');
  }

  console.info({ msg: 'executing_migration_chain', chain: fileChain });

  await db.runInTranscation(async (tx) => {
    console.info({ msg: 'recheking_current_version' });
    const version = await db.getLastVersion(tx);
    if (version !== currentVersion) {
      console.error({
        msg: 'current_version_recheck_failed',
        expectedVersion: currentVersion,
        version,
      });
      throw new Error('Bad version');
    }

    // eslint-disable-next-line no-loops/no-loops
    for (let i = 0; i < fileChain.length; i += 1) {
      const fileName = fileChain[i];
      const filePath = migrationFiles.migrationFilePaths[fileName];
      console.info({
        msg: 'executing_migration_step',
        fileName,
        filePath,
      });
      try {
        // eslint-disable-next-line no-await-in-loop
        await pTimeout(
          db.runQueryFromFile(filePath, tx),
          MIGRATION_STEP_TIMEOUT_MS,
        );
      } catch (error) {
        console.error({
          msg: 'migration_step_failed',
          fileName,
          // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
          error_message: error.message,
          // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
          error_stack: error.stack,
        });
        throw new Error('Migration step failed');
      }
    }

    console.info({ msg: 'adding_migration_record' });

    await db.addMigration(
      {
        from_version: version,
        to_version: targetVersion,
        instance_id: 'inst_id',
        service_version: '000',
      },
      tx,
    );
  });
  console.info({ msg: 'migration_complete', targetVersion });
};


src/migrate/MigrationFiles.ts
import path from 'path';
import fs from 'fs';

import { Graph, alg } from 'graphlib';

export default class MigrationFiles {
  migrationFilePaths: Record<string, string> = {};
  versions: Array<string> = [];
  versionToVersions: Record<string, Array<string>> = {};
  versionsGraph: Graph;
  paths: Record<string, Record<string, Array<string>>> = {};
  latestVersion: string = '000';

  constructor(directoryPath: string) {
    this.readFiles(directoryPath);
    this.versionsGraph = new Graph({ directed: true });
    this.calcGraph();
    this.calcShortestPaths();
  }

  readFiles(directoryPath: string): void {
    const entries = fs.readdirSync(directoryPath);
    entries.forEach((fileName) => {
      const filePath = path.resolve(directoryPath, fileName);
      const stat = fs.lstatSync(filePath);
      if (!stat.isFile) {
        return;
      }
      const result = /^(\d{3})_to_(\d{3})\.sql$/.exec(fileName);
      if (!result) {
        return;
      }
      const from_version = result[1];
      const to_version = result[2];

      if (!this.versions.includes(from_version)) {
        this.versions.push(from_version);
      }
      if (!this.versionToVersions[from_version]) {
        this.versionToVersions[from_version] = [];
      }
      if (!this.versions.includes(to_version)) {
        this.versions.push(to_version);
      }
      if (!this.versionToVersions[to_version]) {
        this.versionToVersions[to_version] = [];
      }
      if (!this.versionToVersions[from_version].includes(to_version)) {
        this.versionToVersions[from_version].push(to_version);
      }
      this.migrationFilePaths[fileName] = filePath;
    });
  }

  calcGraph(): void {
    this.versions.forEach((version) => {
      this.versionsGraph.setNode(version);
      if (this.latestVersion < version) {
        this.latestVersion = version;
      }
    });
    Object.entries(this.versionToVersions).forEach(
      ([fromVersion, toVersions]) => {
        toVersions.forEach((toVersion) => {
          this.versionsGraph.setEdge(fromVersion, toVersion);
        });
      },
    );
  }

  calcShortestPaths(): void {
    const links = alg.dijkstraAll(this.versionsGraph);

    const getFileChain = (
      fromVersion: string,
      toVersion: string,
    ): Array<string> => {
      const link = links[fromVersion][toVersion];
      const { predecessor } = link;
      if (predecessor === fromVersion) {
        return [`${fromVersion}_to_${toVersion}.sql`];
      }
      return [
        ...getFileChain(fromVersion, predecessor),
        `${predecessor}_to_${toVersion}.sql`,
      ];
    };

    this.versions.forEach((fromVersion) => {
      this.versions.forEach((toVersion) => {
        if (fromVersion === toVersion) {
          return;
        }
        if (links[fromVersion][toVersion].distance === Infinity) {
          return;
        }
        const fileChain = getFileChain(fromVersion, toVersion);
        if (!this.paths[fromVersion]) {
          this.paths[fromVersion] = {};
        }
        this.paths[fromVersion][toVersion] = fileChain;
      });
    });
  }
}


src/HttpServer.ts
import express, { Application, Request, Response } from 'express';
import bodyParser from 'body-parser';
import cors from 'cors';

import Db from './Db';

interface HandlerResponseType {
  status: 'OK' | 'ERROR';
}

declare type HandlerType = <TReq, TRsp extends HandlerResponseType>(
  request: TReq,
) => Promise<TRsp>;

export default class HttpServer {
  expressApp: Application;
  port: number;
  db: Db;
  status: 'created' | 'started';

  constructor({ port, db }: { port: number; db: Db }) {
    this.port = port;
    this.db = db;
    this.status = 'created';

    this.expressApp = express();
    this.expressApp.use(cors({
      origin: 'http://localhost:3000',
      optionsSuccessStatus: 200 // For legacy browser support
    }));
    this.expressApp.use(bodyParser.json());
    this.expressApp.use(bodyParser.urlencoded({ extended: false }));

    this.expressApp.post(
      '/persons/add',
      async (req: Request, res: Response) => {
        // persons.push({...req.body, id: uuidv4()})
        await this.db.addPerson({
          person_name: req.body.name,
          birthday: req.body.birthday,
        });
        res.sendStatus(200);
      },
    );

    this.expressApp.post(
      '/persons/flush',
      async (req: Request, res: Response) => {
        await this.db.flushPersons();
        res.sendStatus(200);
      },
    );

    this.expressApp.get('/persons', async (req: Request, res: Response) => {
      const personList = await this.db.personsList({ skip: 0, limit: 1000 });
      res.send(personList);
    });

    this.expressApp.get('/persons/:uuid', (req: Request, res: Response) => {
      res.send({});
    });
  }

  async start(): Promise<void> {
    if (this.status === 'started') {
      throw new Error('Already started');
    }

    return new Promise((resolve, reject) => {
      // @ts-ignore
      this.expressApp.listen(this.port, '0.0.0.0', (error) => {
        if (error) {
          reject(error);
          return;
        }
        console.log(`app started ${this.port}`);
        resolve();
      });
      this.status = 'started';
    });
  }
}
